File: src/open_second_brain/install_cli.py
"""Install (and optionally remove) CLI symlinks for o2b and vault-log.

After ``hermes plugins install`` (or ``pip install``), the ``o2b`` and
``vault-log`` entry-points may not be on PATH. This module creates
symlinks from ``~/.local/bin/o2b`` → ``scripts/o2b`` (and the same for
vault-log) so the bare commands work immediately.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from pathlib import Path

CLI_SCRIPTS: tuple[str, ...] = ("o2b", "vault-log")


@dataclass(frozen=True)
class InstallResult:
    """Outcome of an install-cli run."""

    bindir: Path
    outcomes: tuple[tuple[str, str], ...]
    errors: tuple[str, ...] = field(default_factory=tuple)


@dataclass(frozen=True)
class UninstallResult:
    """Outcome of a remove-cli run."""

    bindir: Path
    outcomes: tuple[tuple[str, str], ...]
    errors: tuple[str, ...] = field(default_factory=tuple)


def _repo_root() -> Path:
    """Return the repository root (two directories above this module)."""
    return Path(__file__).resolve().parent.parent.parent


def _scripts_dir() -> Path:
    """Return the ``scripts/`` directory inside the repo root."""
    return _repo_root() / "scripts"


def _find_script(name: str) -> Path | None:
    """Return the absolute path to a script, or ``None`` if it is missing."""
    script = _scripts_dir() / name
    if script.is_file():
        return script.resolve()
    return None


def _is_valid_symlink(link: Path, target: Path) -> bool:
    """Return ``True`` when *link* points to *target*."""
    try:
        return link.resolve() == target.resolve()
    except OSError:
        return False


def install_cli(bindir: Path | None = None) -> InstallResult:
    """Create (or verify) symlinks for o2b and vault-log.

    Args:
        bindir: Target directory. Defaults to ``~/.local/bin``.

    Returns an :class:`InstallResult`.
    """
    if bindir is None:
        bindir = Path.home() / ".local" / "bin"

    bindir.mkdir(parents=True, exist_ok=True)

    outcomes: list[tuple[str, str]] = []
    errors: list[str] = []

    for name in CLI_SCRIPTS:
        link = bindir / name
        source = _find_script(name)

        if source is None:
            msg = f"error: script 'scripts/{name}' not found in {_scripts_dir()}"
            outcomes.append((name, msg))
            errors.append(msg)
            continue

        if link.is_symlink():
            if _is_valid_symlink(link, source):
                outcomes.append((name, f"exists: {link} → {source}"))
            else:
                existing_target = "unknown"
                try:
                    existing_target = str(link.readlink())
                except OSError:
                    pass
                msg = f"warning: {link} already points to {existing_target}, not overwriting"
                outcomes.append((name, msg))
        elif link.exists():
            msg = f"warning: {link} exists and is not a symlink, not overwriting"
            outcomes.append((name, msg))
        else:
            try:
                link.symlink_to(source)
                outcomes.append((name, f"created: {link} → {source}"))
            except OSError as exc:
                msg = f"error: could not create symlink {link}: {exc}"
                outcomes.append((name, msg))
                errors.append(msg)

    return InstallResult(
        bindir=bindir,
        outcomes=tuple(outcomes),
        errors=tuple(errors),
    )


def uninstall_cli(bindir: Path | None = None) -> UninstallResult:
    """Remove the symlinks created by :func:`install_cli`.

    Only removes a symlink when it points into the ``scripts/`` directory
    of **this** repository (to avoid removing unrelated symlinks that happen
    to share the same name).

    Args:
        bindir: Directory with the symlinks. Defaults to ``~/.local/bin``.

    Returns an :class:`UninstallResult`.
    """
    if bindir is None:
        bindir = Path.home() / ".local" / "bin"

    repo_scripts = _scripts_dir()
    outcomes: list[tuple[str, str]] = []
    errors: list[str] = []

    for name in CLI_SCRIPTS:
        link = bindir / name

        if not link.is_symlink():
            if link.exists():
                msg = f"skipped: {link} is not a symlink — refusing to remove"
                outcomes.append((name, msg))
            else:
                outcomes.append((name, f"skipped: {link} does not exist"))
            continue

        # Resolve the target of the symlink
        try:
            target = link.resolve()
        except OSError as exc:
            msg = f"error: cannot resolve {link}: {exc}"
            outcomes.append((name, msg))
            errors.append(msg)
            continue

        # Only remove if the target is inside our scripts/ directory
        try:
            target.relative_to(repo_scripts)
        except ValueError:
            msg = f"skipped: {link} → {target} is outside this repo's scripts/ — refusing to remove"
            outcomes.append((name, msg))
            continue

        try:
            link.unlink()
            outcomes.append((name, f"removed: {link}"))
        except OSError as exc:
            msg = f"error: cannot unlink {link}: {exc}"
            outcomes.append((name, msg))
            errors.append(msg)

    return UninstallResult(
        bindir=bindir,
        outcomes=tuple(outcomes),
        errors=tuple(errors),
    )


def render_install_result(result: InstallResult) -> str:
    """Render an :class:`InstallResult` as plain text."""
    lines: list[str] = []
    lines.append(f"o2b install-cli — {result.bindir}")
    lines.append("-" * 40)
    for name, msg in result.outcomes:
        lines.append(f"  {name}: {msg}")
    if result.errors:
        lines.append("")
        lines.append(f"{len(result.errors)} error(s).")
    return "\n".join(lines).rstrip() + "\n"


def render_uninstall_result(result: UninstallResult) -> str:
    """Render an :class:`UninstallResult` as plain text."""
    lines: list[str] = []
    lines.append(f"o2b uninstall --remove-cli — {result.bindir}")
    lines.append("-" * 40)
    for name, msg in result.outcomes:
        lines.append(f"  {name}: {msg}")
    if result.errors:
        lines.append("")
        lines.append(f"{len(result.errors)} error(s).")
    return "\n".join(lines).rstrip() + "\n"

File: src/cli/install-cli.ts
/**
 * Install (and remove) CLI symlinks for `o2b` and `vault-log` in `~/.local/bin`.
 *
 * Mirrors `src/open_second_brain/install_cli.py`. Refuses to overwrite a
 * symlink that already points to a different repo's checkout — that's the
 * documented behavior across multi-runtime installs.
 */

import {
  existsSync,
  lstatSync,
  mkdirSync,
  readlinkSync,
  realpathSync,
  statSync,
  symlinkSync,
  unlinkSync,
} from "node:fs";
import { homedir } from "node:os";
import { dirname, join, resolve, sep } from "node:path";
import { fileURLToPath } from "node:url";

import type { InstallResult, UninstallResult } from "../core/types.ts";

const CLI_SCRIPTS = ["o2b", "vault-log"] as const;

function repoRoot(): string {
  // src/cli/install-cli.ts → repo/
  return resolve(dirname(fileURLToPath(import.meta.url)), "..", "..");
}

function scriptsDir(): string {
  return join(repoRoot(), "scripts");
}

function findScript(name: string): string | null {
  const path = join(scriptsDir(), name);
  if (existsSync(path) && statSync(path).isFile()) return resolve(path);
  return null;
}

function isLink(p: string): boolean {
  try {
    return lstatSync(p).isSymbolicLink();
  } catch {
    return false;
  }
}

/**
 * Match Python's `Path.resolve()` semantics, which follows symlinks. Plain
 * `path.resolve()` does not — it just normalises path components.
 */
function realpath(p: string): string {
  try {
    return realpathSync(p);
  } catch {
    return resolve(p);
  }
}

function isValidSymlink(link: string, target: string): boolean {
  try {
    return realpath(link) === realpath(target);
  } catch {
    return false;
  }
}

export function installCli(bindir?: string): InstallResult {
  const dir = bindir ?? join(homedir(), ".local", "bin");
  mkdirSync(dir, { recursive: true });

  const outcomes: Array<readonly [string, string]> = [];
  const errors: string[] = [];

  for (const name of CLI_SCRIPTS) {
    const link = join(dir, name);
    const source = findScript(name);
    if (source === null) {
      const msg = `error: script 'scripts/${name}' not found in ${scriptsDir()}`;
      outcomes.push([name, msg]);
      errors.push(msg);
      continue;
    }

    if (isLink(link)) {
      if (isValidSymlink(link, source)) {
        outcomes.push([name, `exists: ${link} → ${source}`]);
      } else {
        let existing = "unknown";
        try {
          existing = readlinkSync(link);
        } catch {
          // ignore
        }
        outcomes.push([
          name,
          `warning: ${link} already points to ${existing}, not overwriting`,
        ]);
      }
    } else if (existsSync(link)) {
      outcomes.push([name, `warning: ${link} exists and is not a symlink, not overwriting`]);
    } else {
      try {
        symlinkSync(source, link);
        outcomes.push([name, `created: ${link} → ${source}`]);
      } catch (exc) {
        const msg = `error: could not create symlink ${link}: ${(exc as Error).message ?? exc}`;
        outcomes.push([name, msg]);
        errors.push(msg);
      }
    }
  }
  return { bindir: dir, outcomes, errors };
}

export function uninstallCli(bindir?: string): UninstallResult {
  const dir = bindir ?? join(homedir(), ".local", "bin");
  const repoScripts = scriptsDir();
  const outcomes: Array<readonly [string, string]> = [];
  const errors: string[] = [];

  for (const name of CLI_SCRIPTS) {
    const link = join(dir, name);
    if (!isLink(link)) {
      if (existsSync(link)) {
        outcomes.push([name, `skipped: ${link} is not a symlink — refusing to remove`]);
      } else {
        outcomes.push([name, `skipped: ${link} does not exist`]);
      }
      continue;
    }

    let target: string;
    try {
      target = realpath(link);
    } catch (exc) {
      const msg = `error: cannot resolve ${link}: ${(exc as Error).message ?? exc}`;
      outcomes.push([name, msg]);
      errors.push(msg);
      continue;
    }

    const repoScriptsReal = realpath(repoScripts);
    if (!target.startsWith(repoScriptsReal + sep) && target !== repoScriptsReal) {
      outcomes.push([
        name,
        `skipped: ${link} → ${target} is outside this repo's scripts/ — refusing to remove`,
      ]);
      continue;
    }

    try {
      unlinkSync(link);
      outcomes.push([name, `removed: ${link}`]);
    } catch (exc) {
      const msg = `error: cannot unlink ${link}: ${(exc as Error).message ?? exc}`;
      outcomes.push([name, msg]);
      errors.push(msg);
    }
  }
  return { bindir: dir, outcomes, errors };
}

export function renderInstallResult(result: InstallResult): string {
  const lines: string[] = [];
  lines.push(`o2b install-cli — ${result.bindir}`);
  lines.push("-".repeat(40));
  for (const [name, msg] of result.outcomes) {
    lines.push(`  ${name}: ${msg}`);
  }
  if (result.errors.length > 0) {
    lines.push("");
    lines.push(`${result.errors.length} error(s).`);
  }
  return lines.join("\n").replace(/\s+$/g, "") + "\n";
}

export function renderUninstallResult(result: UninstallResult): string {
  const lines: string[] = [];
  lines.push(`o2b uninstall --remove-cli — ${result.bindir}`);
  lines.push("-".repeat(40));
  for (const [name, msg] of result.outcomes) {
    lines.push(`  ${name}: ${msg}`);
  }
  if (result.errors.length > 0) {
    lines.push("");
    lines.push(`${result.errors.length} error(s).`);
  }
  return lines.join("\n").replace(/\s+$/g, "") + "\n";
}
